### PR TITLE
format detection for vaapi+wayland : accept/reject based on compositor support

### PR DESCRIPTION
**Pull request by boxerab**
For formats that are **not** supported by hardware, this PR enables the choice of an upload format that is supported by the compositor.

### CODE REVIEW

**Comment by boxerab** [comment]
cc @philipl

**Comment by boxerab** [comment]
@philipl thanks, I have moved the wl-specific code into the wldmabuf ra.

**Comment by boxerab** [comment]
[jellyfish-3-mbps-hd-hevc-10bit.zip](https://github.com/mpv-player/mpv/files/10548434/jellyfish-3-mbps-hd-hevc-10bit.zip)

@philipl unfortunately, this `P10` video will not play properly on my system, even with this patch. P10 decoding is supported on my card, while P10 display is not, but mpv still tries hw decoding, which then fails to display.

**Review by philipl — CHANGES_REQUESTED** [review]
Thanks for continuing to work on this.

I've made a suggestion on how to better encapsulate the check.

As for P010, and to be honest, NV12, the actual logic that decides what we can decode (as opposed to what we can present) is completely separate from anything here. This already has the consequence that if a decoder produces an unpresentable format (which you can repro using normal VOs by decoding 4:2:2 hevc content on Intel), it will decode successfully, fail to present, and then just keep failing frame after frame.

Ideally, we would detect this case and use vaapi format conversion (like we do when you upload a supported but unpresentable format) but that's net new functionality.

So anyway, this is an existing problem that should be solvable in a way that's not wl_dmabuf specific.

File: video/out/gpu/ra.h
Comment (boxerab) [reply video/out/gpu/ra.h:516]: But yes, I can move this callback to `ra_wldmabuf` only

Comment (boxerab) [reply video/out/gpu/ra.h]: Done.

File: video/out/hwdec/hwdec_vaapi.c
Comment (philipl) [inline video/out/hwdec/hwdec_vaapi.c:114]: Use MP_TARRAY_APPEND rather than hard coding an array size here. (This is how `formats` above is handled)

File: video/out/wldmabuf/ra_wldmabuf.c
Comment (philipl) [inline video/out/wldmabuf/ra_wldmabuf.c:37]: Where does the magic 16 come from? Make it a named constant.

**Review by philipl — COMMENTED** [review]
Thanks for iterating on this. Just a few small comments left.

File: video/out/hwdec/hwdec_vaapi.c
Comment (boxerab) [reply video/out/hwdec/hwdec_vaapi.c]: fixed.

File: video/out/wldmabuf/ra_wldmabuf.c
Comment (boxerab) [reply video/out/wldmabuf/ra_wldmabuf.c]: It's the size of the struct that stores format info. fixed.

File: video/out/hwdec/hwdec_vaapi.c
Comment (boxerab) [reply video/out/hwdec/hwdec_vaapi.c]: thx, fixed.

**Review by philipl — APPROVED** [review]
Great. Thanks!

File: video/out/hwdec/hwdec_vaapi.c
Comment (sfan5) [inline video/out/hwdec/hwdec_vaapi.c:549]: 1) there is no guarantee that supported_wl_formats is as long as formats
2) most people do not use wldmabuf so this will just print useless and potentially incorrect data

File: video/out/hwdec/hwdec_vaapi.c
Comment (Dudemanguy) [inline video/out/hwdec/hwdec_vaapi.c:296]: I'm not sure how exactly this would be be done, but I feel like hitting this error condition should propagate to dmabuf-wayland and cause it to error out. Or well if you know that you can't map any surfaces at all, then there's no point in continuing.

File: video/out/hwdec/hwdec_vaapi.c
Comment (boxerab) [reply video/out/hwdec/hwdec_vaapi.c]: Good point - fixed.

File: video/out/hwdec/hwdec_vaapi.c
Comment (boxerab) [reply video/out/hwdec/hwdec_vaapi.c:296]: This error condition is normal - it just means that this particular surface is not supported.

Comment (boxerab) [reply video/out/hwdec/hwdec_vaapi.c:296]: Yes, if no surfaces can be mapped, and the ra is wldmabuf, then we should error out.

Comment (boxerab) [reply video/out/hwdec/hwdec_vaapi.c:298]: I've added this sanity check.

File: video/out/hwdec/hwdec_vaapi.c
Comment (sfan5) [inline video/out/hwdec/hwdec_vaapi.c:296]: what about the suggestion to move this to `try_format_pixfmt`?

Comment (boxerab) [reply video/out/hwdec/hwdec_vaapi.c]: This code needs the mapper, so best I could do was move it into `try_format_map`.

**Comment by Dudemanguy** [comment]
This currently fails if you disable wayland and enable vaapi. Or well more accurately, if `dmabuf-wayland` isn't available. Could just guard the functions, but it gets ugly with all the `ra_is_wldmabuf` calls. Not sure if there's something better.

**Comment by boxerab** [comment]
> This currently fails if you disable wayland and enable vaapi. Or well more accurately, if `dmabuf-wayland` isn't available. Could just guard the functions, but it gets ugly with all the `ra_is_wldmabuf` calls. Not sure if there's something better.

Thanks for checking - I added some guards and a small refactor - should be good now

**Comment by Dudemanguy** [comment]
Hmm CI fails for some reason although it doesn't for me locally.

**Comment by boxerab** [comment]
> Hmm CI fails for some reason although it doesn't for me locally.

I had to add a few more guards. Should build fine now.

File: video/out/hwdec/hwdec_vaapi.c
Comment (Dudemanguy) [inline video/out/hwdec/hwdec_vaapi.c:92]: Should be `HAVE_DMABUF_WAYLAND` (same for the other one).

Comment (boxerab) [reply video/out/hwdec/hwdec_vaapi.c]: yes, fixed now.

**Review by Dudemanguy — APPROVED** [review]
I think this is good now.

**Comment by boxerab** [comment]
@Dudemanguy any more issues with the PR ?

**Comment by Dudemanguy** [comment]
Just waiting on another review.

**Comment by philipl** [comment]
@sfan5 had thoughts.

**Comment by sfan5** [comment]
I am still unhappy that the changes to `hwdec_vaapi.c` break the abstraction by adding highly wldmabuf-specific code paths in there (and kind of surprised neither of you seem to mind).
For example this code can reasonably be moved to `map` in dmabuf_interop_wl.c:
grafik
A different solution has to be found for toggling `VA_EXPORT_SURFACE_COMPOSED_LAYERS` but that's not hard either (add a flag to `struct dmabuf_interop`).

By the way, what is the point of collecting the formats into `supported_wl_formats`? All you do is check whether the list is non-empty once at startup.

<hr>

As an aside I just went checking whether any comment message mentions why dmabuf_interop_wl is all stubs (there is one) and stumbled upon a3340645beade8a92eae90ac89c467e1cb439741 and at a glance:

wonder what happens when it's full
grafik

locking was removed with no mention _anywhere_ why it's now apparently superfluous
grafik

Am I being overly pedantic or does nobody properly review these?

**Comment by philipl** [comment]
@sfan5 You make a good point on using `dmabuf_interop_wl`. Honestly, I forgot it gave us a place to stand to do this.

I can't comment on the second change, as I never looked at that one.

**Comment by Dudemanguy** [comment]
@sfan5: Sorry, it's not obvious from the commit alone and I didn't have him explain it and I guess I should have, but locking isn't needed anymore. The reason is because the buffer pool is now separated into two different lists that are consistent and don't unpredictably change. Beforehand, it was a mixture of available and pending free buffers. Since the compositor could free a buffer at any time, a lock was needed. With the newer change, the compositor only operates on the pending list and mpv won't possibly try to allocate a new entry while something is being freed.

If `WLBUF_NUM_PURG_ENTRIES` is full, it will simply leak that particular buffer but that situation should basically never happen. From my experience, compositors typically only hold onto a handful of buffers at a time. There's basically no reason why it should ever go anywhere near 60 in the first place which is way overkill. I didn't think writing a failcase for that was necessary, so I didn't ask. The actual number is just `WLBUF_POOL_NUM_ALLOCATED_INIT` * 2. `num_allocated` will dynamically reallocate itself (double in size) if the pool of available images fills up but this is, again, something that should never happen.

**Comment by boxerab** [comment]
@sfan5 thanks for your review - I have moved all wldmabuf code into the wl interop, so hwdec_vaapi is now almost unchanged. 

As for locking, this was never needed as pool is always accessed from the vo thread. But yes, this should have been commented on in the commit message.

File: video/out/hwdec/hwdec_vaapi.c
Comment (philipl) [inline video/out/hwdec/hwdec_vaapi.c:305]: Why the change here? Because you noticed `GRAY8` has zero planes? I don't know what's up with that, but I don't think it's wrong to reject a malformed format like that.

Comment (boxerab) [reply video/out/hwdec/hwdec_vaapi.c:305]: This was added to differentiate `dmabuf _interop_wl` from others, as this interop doesn't need to set the number of planes. In this case, we accept formats based purely on what Wayland reports it supports, so we ignore this check on returned planes.

File: video/out/hwdec/hwdec_vaapi.c
Comment (boxerab) [reply video/out/hwdec/hwdec_vaapi.c]: This has been moved into the dmabuf interop.

**Comment by sfan5** [comment]
Here are my suggested changes: http://sprunge.us/oeadrU?diff

**Comment by boxerab** [comment]
> Here are my suggested changes: http://sprunge.us/oeadrU?diff

Thanks, can you attach the diff file to the issue ?

**Comment by sfan5** [comment]
You can retrieve the raw text by removing the `?diff`.

> If `WLBUF_NUM_PURG_ENTRIES` is full, it will simply leak that particular buffer but that situation should basically never happen.

Okay cool, but a code comment should explain this.

> I didn't think writing a failcase for that was necessary, so I didn't ask.

Writing a dynamic implementation with `MP_TARRAY_APPEND` would have been negligibly more work, but instead we get this hardcoded magic number.
Since you say it only holds a handful of buffers I can even think of a concrete advantage: a linear search on 60 elements usually takes more time.

**Comment by boxerab** [comment]
> You can retrieve the raw text by removing the `?diff`.

thx, I had to make a few small changes, but this is now pushed.

**Review by sfan5 — APPROVED** [review]
Code looks OK

**Comment by boxerab** [comment]
@philipl @sfan5 do you have any other concerns about this PR ?

File: video/out/wayland_common.c
Comment (Dudemanguy) [inline video/out/wayland_common.c:2237]: You need to remove this from `wayland_common.h` as well.

File: video/out/hwdec/dmabuf_interop_wl.c
Comment (Dudemanguy) [inline video/out/hwdec/dmabuf_interop_wl.c:74]: Any reason to not just use `strcmp` here? I know this annoying used to be called `vaapi-egl` for a long time, but I'm pretty sure the name won't be changing anytime soon.

Comment (boxerab) [reply video/out/hwdec/dmabuf_interop_wl.c:74]: > I'm pretty sure the name won't be changing anytime soon

famous last words :)

File: video/out/wldmabuf/ra_wldmabuf.c
Comment (Dudemanguy) [inline video/out/wldmabuf/ra_wldmabuf.c:50]:
```suggestion
struct ra *ra_create_wayland(struct mp_log *log, struct vo *vo)
```

Comment (boxerab) [reply video/out/wldmabuf/ra_wldmabuf.c]: fixed

File: video/out/hwdec/dmabuf_interop_wl.c
Comment (boxerab) [reply video/out/hwdec/dmabuf_interop_wl.c]: fixed

File: video/out/wayland_common.c
Comment (boxerab) [reply video/out/wayland_common.c:2237]: done

**Review by Dudemanguy — APPROVED** [review]
Okay I think this time we're good.

File: video/out/hwdec/dmabuf_interop_wl.c
Comment (jeeb) [inline video/out/hwdec/dmabuf_interop_wl.c:45]: Not sure if this is actually possible to be at zero or less, but would actually seem like a better way to first check, then utilize?  I see desc being an AVDRMFrameDescriptor which has a static amount of planes and formats always available so positively enough even without checking there should be no out of invalid memory access.

Not sure about the message level. Unless this is spammy, maybe follow other probing components and do `MP_MSG` with `probing ? MSGL_V : MSGL_ERR`?

```c
    // we're only reading things from this structure so const should be fine
    const AVDRMFrameDescriptor desc =
        ((struct dmabuf_interop_priv *)mapper->priv)->desc;

    if (desc.nb_layers != 1 || desc.nb_objects != 1) {
        MP_DBG(mapper, "Mapping attempted with surface which has unexpected "
                       "amount of layers (%s, %d) or objects (%s, %d)!\n",
               desc.nb_layers != 1 ? "yes" : "no",
               desc.nb_layers,
               desc.nb_objects != 1 ? "yes" : "no",
               desc.nb_objects);
        return false;
    }

    uint32_t drm_format      = desc.layers[0].format;
    uint64_t format_modifier = desc.objects[0].format_modifier;

    if (!ra_compatible_format(mapper->ra, drm_format, format_modifier) {
        MP_DBG(mapper, "Mapping attempted with an unsupported surface with "
                       "format %s; drm format '%s(%016lx)'!\n",
                       mp_imgfmt_to_name(mapper->src->params.hw_subfmt),
                       mp_tag_str(drm_format),
                       format_modifier);
        return false;
    }

    MP_DBG(mapper, "Mapping supported Wayland display format: '%s(%016lx)'\n",
           mp_tag_str(drm_format), format_modifier);
```

Comment (Dudemanguy) [reply video/out/hwdec/dmabuf_interop_wl.c:45]: Made it always verbose.

**Comment by Dudemanguy** [comment]
@boxerab: I just changed some `MP_DBG` to `MP_VERBOSE` in case you were wondering.

**Pull request merged**
